packages/luigis-box/src/Model/Batch/LuigisBoxBatchLoader.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Shopsys\LuigisBoxBundle\Model\Batch;
+
+use GraphQL\Executor\Promise\Promise;
+use GraphQL\Executor\Promise\PromiseAdapter;
+use Shopsys\FrameworkBundle\Model\Product\Search\FilterQueryFactory;
+use Shopsys\FrameworkBundle\Model\Product\Search\ProductElasticsearchRepository;
+use Shopsys\LuigisBoxBundle\Component\LuigisBox\LuigisBoxClient;
+use Shopsys\LuigisBoxBundle\Component\LuigisBox\LuigisBoxResult;
+
+class LuigisBoxBatchLoader
+{
+    /**
+     * @var array<string, int>
+     */
+    protected static array $totalsByType = [];
+
+    /**
+     * @param \Shopsys\LuigisBoxBundle\Component\LuigisBox\LuigisBoxClient $luigisBoxClient
+     * @param \GraphQL\Executor\Promise\PromiseAdapter $promiseAdapter
+     * @param \Shopsys\FrameworkBundle\Model\Product\Search\ProductElasticsearchRepository $productElasticsearchRepository
+     * @param \Shopsys\FrameworkBundle\Model\Product\Search\FilterQueryFactory $filterQueryFactory
+     */
+    public function __construct(
+        protected readonly LuigisBoxClient $luigisBoxClient,
+        protected readonly PromiseAdapter $promiseAdapter,
+        protected readonly ProductElasticsearchRepository $productElasticsearchRepository,
+        protected readonly FilterQueryFactory $filterQueryFactory,
+    ) {
+    }
+
+    /**
+     * @param string $type
+     * @return int
+     */
+    public static function getTotalByType(string $type): int
+    {
+        return self::$totalsByType[$type] ?? 0;
+    }
+
+    /**
+     * @param \Shopsys\LuigisBoxBundle\Model\Batch\LuigisBoxBatchLoadData[] $luigisBoxBatchLoadData
+     * @return \GraphQL\Executor\Promise\Promise
+     */
+    public function loadByBatchData(array $luigisBoxBatchLoadData): Promise
+    {
+        $query = '';
+        $limitsByType = [];
+        $endpoint = '';
+        $page = 0;
+        $filter = [];
+        $orderingMode = null;
+
+        foreach ($luigisBoxBatchLoadData as $luigisBoxBatchLoadDataItem) {
+            if ($luigisBoxBatchLoadDataItem->getType() === LuigisBoxClient::MAIN_TYPE) {
+                $query = $luigisBoxBatchLoadDataItem->getQuery();
+                $endpoint = $luigisBoxBatchLoadDataItem->getEndpoint();
+                $page = $luigisBoxBatchLoadDataItem->getPage();
+                $filter = $luigisBoxBatchLoadDataItem->getFilter();
+                $orderingMode = $luigisBoxBatchLoadDataItem->getOrderingMode();
+            }
+
+            $limitsByType[$luigisBoxBatchLoadDataItem->getType()] = $luigisBoxBatchLoadDataItem->getLimit();
+        }
+
+        return $this->promiseAdapter->all($this->mapDataByTypes(
+            $this->luigisBoxClient->getData($query, $endpoint, $page, $limitsByType, $filter, $orderingMode),
+            $limitsByType,
```

Comment:
This seems confusing to me... limitsByType is widely used to determine the proper search. I'm not sure how the filter and orderingMode fit into that.
Is it exclusive to products? Can it be set differently for products and categories? 

I would rather have something like `getData($query, $isAutocomplete, $batchLoadDataByTypes)`

deciding the endpoint at the top level seems like a violation of encapsulation... the proper endpoint name should be the responsibility of the client class itself. The input data should be only if I want to fullsearch or autocomplete.

passing the array of configurations is more open to changes because I'm able to easily add for example independent pagination for categories and products.

Reply:
I have renamed used variables to be more clear, I have added validation for endpoints and also renamed the endpoint to action and we have discussed configuration and decided to keep it as it is.